README.md
# tree_structures

tree_structures is a Dart implementation of (currently only) a red-black self-balancing binary search tree in Dart with a possibility to output to Graphviz for previewing.
Red-black tree solution is based on the logic and structure of the [Franck Bui-Huu's C implementation](https://github.com/fbuihuu/libtree/blob/master/rb.c).

Because of the pure Dart solution, it can't really reach high speeds, so the usage is simply educational.
The implementation focuses on correctness, providing assertion for conformance to the red-black tree rules during testing.
It also takes advantage of the Sound null safety.

The tree is generic, accepting `K extends Comparable` for the key, and any `V` for the value types.

Right now the benchmark is as following (with comparison to a regular `Map`), using `int` as keys:

## Count-based benchmark

```
count: 500000

adding random to a regular [map]: 116ms

adding random to [randomTree]: 309ms
testing conformity of [randomTree]: 71ms
searching in [randomTree]: 227ms

adding sequential to [sequentialTree]: 242ms
testing conformity of [sequentialTree]: 35ms
successor & predecessor in [sequentialTree]: 58ms
deletion in [sequentialTree]: 81ms
```

## Time-based benchmark

```
finished with 1328915 loops (332357 inserts, 996558 skipped inserts, 332272 deletions) within 5000ms, keys in [0..100) range.
```

## Keys output

```dart
var tree = RBTree<int, dynamic>();
tree.insert(100, "data associated with 100");
tree.insert(150, "data associated with 150");
tree.remove(100);
tree.insert(200, "data associated with 200");
tree.insert(20, "data associated with 20");
tree.insert(25, "data associated with 25");
tree.insert(18, "data associated with 18");
tree.insert(120, "data associated with 120");
tree.insert(180, "data associated with 180");
```

Output:
```
(150(20<r(18<b)(25>b(120>r)))(200>b(180<r)))
```

## `forEach`

```dart
tree.forEach((node) {
  print("${node.key}:${node.value}");
  return true; // continue
});
```

Output:
```
18:data associated with 18
20:data associated with 20
25:data associated with 25
120:data associated with 120
150:data associated with 150
180:data associated with 180
200:data associated with 200
```

## Output to `Graphviz`

```dart
print(tree.output(OutputStyle.Graphviz));
```

Output:
```
digraph {
  150[style=filled,color=black,fontcolor=white];
  150 -> 20;
  20[style=filled,color=tomato,fontcolor=white];
  20 -> 18;
  18[style=filled,color=black,fontcolor=white];
  null0 [shape=point];
  18 -> null0;
  null1 [shape=point];
  18 -> null1;
  20 -> 25;
  25[style=filled,color=black,fontcolor=white];
  null2 [shape=point];
  25 -> null2;
  25 -> 120;
  120[style=filled,color=tomato,fontcolor=white];
  null3 [shape=point];
  120 -> null3;
  null4 [shape=point];
  120 -> null4;
  150 -> 200;
  200[style=filled,color=black,fontcolor=white];
  200 -> 180;
  180[style=filled,color=tomato,fontcolor=white];
  null5 [shape=point];
  180 -> null5;
  null6 [shape=point];
  180 -> null6;
  null7 [shape=point];
  200 -> null7;
}
```

Graphviz to svg example output of the above tree (`dot -Tsvg <file.dot> > exmple_output.svg`):

![svg](https://raw.githubusercontent.com/iuthere/tree_structures/master/example_output.svg)

## pub.dev

https://pub.dev/packages/tree_structures

CHANGELOG.md
## 0.0.5

* Provide an example, add more API docs.

## 0.0.4

* README link back to pub.dev.

## 0.0.3

* README links fix.

## 0.0.2

* Repository addition.

## 0.0.1

* Initial solution.

test/red_black_tree_test.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:tree_structures/tree_structures.dart';

extension on Stopwatch {
  restart() {
    reset();
    start();
  }

  reportMs(String message) {
    stop();
    stdout.write("$message: ${elapsed.inMilliseconds}ms\n");
    reset();
  }
}

void main() {
  test("RBTree Expectation", () {
    var tree = RBTree<int, dynamic>();
    var _e = (String v, int size) {
      tree.assertConformity();
      expect(tree.output(OutputStyle.KeyDesc).toString(), v);
      expect(tree.length, size);
    };
    _e("", 0);
    expect(tree.first, null);
    expect(tree.last, null);
    tree.insert(100, "");
    _e("(100)", 1);
    tree.remove(100);
    _e("", 0);
    expect(tree.root, null);
    tree.insert(100, "");
    tree.insert(150, "");
    _e("(100(150>r))", 2);
    tree.insert(90, "");
    _e("(100(90<r)(150>r))", 3);
    tree.insert(50, "");
    _e("(100(90<b(50<r))(150>b))", 4);
    tree.remove(90);
    _e("(100(50<b)(150>b))", 3);
    tree.insert(90, "");
    _e("(100(50<b(90>r))(150>b))", 4);
    tree.insert(80, "");
    _e("(100(80<b(50<r)(90>r))(150>b))", 5);
    expect(tree.first, tree.find(50));
    expect(tree.last, tree.find(150));
    expect("100,80,50,90,150", tree.output(OutputStyle.KeyList).toString());
    tree.remove(80);
    _e("(100(90<b(50<r))(150>b))", 4);
    tree.remove(100);
    _e("(90(50<b)(150>b))", 3);
    tree.insert(60, "");
    _e("(90(50<b(60>r))(150>b))", 4);
    tree.insert(45, "");
    _e("(90(50<b(45<r)(60>r))(150>b))", 5);
    tree.insert(25, "");
    _e("(90(50<r(45<b(25<r))(60>b))(150>b))", 6);
    tree.insert(48, "");
    _e("(90(50<r(45<b(25<r)(48>r))(60>b))(150>b))", 7);
    expect(tree.first, tree.find(25));
    expect(tree.last, tree.find(150));

    tree.insert(48, ""); // duplicate
    tree.insert(60, ""); // duplicate

    tree.insert(46, "");
    _e("(50(45<r(25<b)(48>b(46<r)))(90>r(60<b)(150>b)))", 8);
    tree.insert(180, "");
    _e("(50(45<r(25<b)(48>b(46<r)))(90>r(60<b)(150>b(180>r))))", 9);
    tree.insert(190, "");
    _e("(50(45<r(25<b)(48>b(46<r)))(90>r(60<b)(180>b(150<r)(190>r))))", 10);
    expect(tree.first!.key, 25);
    expect(tree.last!.key, 190);

    var sb = StringBuffer();
    tree.forEach((node) {
      if (sb.length > 0) {
        sb.write(",");
      }
      sb.write("${node.key}");
      return true;
    });
    expect(
      sb.toString(),
      "25,45,46,48,50,60,90,150,180,190",
    );
    expect(
      tree.output(OutputStyle.KeyList).toString(),
      "50,45,25,48,46,90,60,180,150,190",
    );

    //outputToDotAndSvg(tree, "test");
    tree.remove(90);
    _e("(50(45<r(25<b)(48>b(46<r)))(150>r(60<b)(180>b(190>r))))", 9);
    tree.remove(48);
    _e("(50(45<r(25<b)(46>b))(150>r(60<b)(180>b(190>r))))", 8);
    tree.remove(50);
    _e("(60(45<r(25<b)(46>b))(180>r(150<b)(190>b)))", 7);
    tree.remove(45);
    _e("(60(46<b(25<r))(180>r(150<b)(190>b)))", 6);
    tree.remove(46);
    expect(tree.first, tree.find(25));
    expect(tree.last, tree.find(190));
    _e("(60(25<b)(180>r(150<b)(190>b)))", 5);
    tree.remove(150);
    _e("(60(25<b)(180>b(190>r)))", 4);
    tree.remove(25);
    _e("(180(60<b)(190>b))", 3);
    expect(tree.first, tree.find(60));
    expect(tree.last, tree.find(190));
    tree.remove(60);
    _e("(180(190>r))", 2);
    tree.remove(190);
    _e("(180)", 1);
    tree.remove(180);
    _e("", 0);
    expect(tree.root, null);
    expect(tree.first, null);
    expect(tree.last, null);
  });
  test("RBTree Performance", () async {
    //var count = 5000000;
    var count = 500000;
    stdout.write("count: $count\n");

    var rnd = Random();

    var timer = Stopwatch();
    timer.restart();
    var map = Map<int, dynamic>();
    for (int i = 0; i < count; i++) {
      var v = rnd.nextInt(count * 10);
      map[v] = v;
      assert(map[v] == v);
    }
    timer.reportMs("adding random to a regular [map]");

    timer.restart();
    var randomTree = RBTree<int, dynamic>();
    map.forEach((key, value) {
      /*var pair = */ randomTree.insert(key, value);
    });
    assert(randomTree.length == map.length);
    // for (int i = 0; i < count; i++) {
    //   var v = rnd.nextInt(count * 10);
    //   var pair = randomTree.insertKV(v, v);
    //   //var found = randomTree.search(randomTree.root!, v);
    //   //assert(found?.value == v);
    //   //assert(found == pair.added);
    // }
    timer.reportMs("adding random to [randomTree]");

    timer.restart();
    randomTree.assertConformity();
    timer.reportMs("testing conformity of [randomTree]");

    timer.restart();

    assert(randomTree.root != null);
    map.forEach((key, value) {
      assert(key == value);
      var f = randomTree.find(key);
      assert(f != null && f.key == key);
    });

    timer.reportMs("searching in [randomTree]");

    timer.restart();
    var sequentialTree = RBTree<int, dynamic>();
    for (int i = 0; i < count; i++) {
      var pair = sequentialTree.insert(i, i);
      var found = sequentialTree.find(i);
      assert(found?.value == i);
      assert(found == pair.inserted);
      assert(sequentialTree.min(sequentialTree.root!).key == 0);
      assert(sequentialTree.max(sequentialTree.root!).key == i);
    }

    timer.reportMs("adding sequential to [sequentialTree]");

    timer.restart();
    sequentialTree.assertConformity();
    timer.reportMs("testing conformity of [sequentialTree]");

    timer.restart();
    for (int i = 0; i < count; i++) {
      var found = sequentialTree.find(i);
      assert(found != null);
      assert(found?.value == i);
      var successor = sequentialTree.successor(found!);
      if (successor == null) {
        assert(i == count - 1);
      } else {
        assert(i + 1 == successor.key);
      }
      var predecessor = sequentialTree.predecessor(found);
      if (predecessor == null) {
        assert(i == 0);
      } else {
        assert(i - 1 == predecessor.key);
      }
    }
    timer.reportMs("successor & predecessor in [sequentialTree]");

    timer.restart();
    for (int i = 0; i < count; i++) {
      var found = sequentialTree.find(i);
      assert(found != null);
      sequentialTree.remove(found!.key);
      if (sequentialTree.root == null) {
        assert(i == count - 1);
      } else {
        found = sequentialTree.find(i);
        assert(found == null);
      }
    }
    timer.reportMs("deletion in [sequentialTree]");
  });
  test("RBTree Random Insert/Delete", () {
    RBTree<int, dynamic> tree = RBTree<int, dynamic>();
    var rnd = Random();
    var timerMs = 5000;
    var range = 100;

    var timer = Stopwatch();
    timer.start();
    var count = 0;
    var operations = 0;
    var inserts = 0;
    var skippedInserts = 0;
    var deletions = 0;
    while (timer.elapsedMilliseconds < timerMs) {
      var key = rnd.nextInt(range);
      var result = tree.insert(key, key);
      if (result.detached == null) {
        inserts++;
        count++;
      } else {
        skippedInserts++;
      }
      expect(tree.length, count);
      tree.assertConformity();

      if (rnd.nextInt(4) < 1) {
        var toDelete = tree.randomNode(rnd);
        if (toDelete != null) {
          tree.remove(toDelete.key);
          deletions++;
          count--;
          expect(tree.length, count);
          tree.assertConformity();
        }
      }
      operations++;
    }
    stdout.write(
      "finished with $operations loops ($inserts inserts, $skippedInserts skipped inserts, $deletions deletions) within ${timerMs}ms, keys in [0..$range) range.\n",
    );

    //print(tree.output(OutputStyle.KeyList));
    //outputToDotAndSvg(tree, "rb_tree_test");
    //outputToFile(tree.marshal(), "test.txt");
    while (tree.length > 0) {
      var toDelete = tree.randomNode(rnd);
      assert(toDelete != null);
      tree.remove(toDelete!.key);
      tree.assertConformity();
    }
    tree.assertConformity();
  });
}

outputToDotAndSvg(RBTree<int, dynamic> tree, String name) async {
  var path = './test/$name.dot';
  var pathSvg = './test/$name.svg';
  var out = File(path).openWrite();
  out.write(tree.output(OutputStyle.Graphviz).toString());
  await out.flush();
  await out.close();
  var result = await Process.run('dot', ['-Tsvg', path]);
  var outSvg = File(pathSvg).openWrite();
  outSvg.write(result.stdout);
  await outSvg.flush();
  await outSvg.close();
}

outputToFile(Uint8List bytes, String name) {
  var path = './test/$name';
  var out = File(path);
  out.openWrite();
  out.writeAsBytes(bytes, flush: true);
}


lib/tree_structures.dart
import 'dart:math';

/// [RBNode] represents a generic node for the red-black binary search tree.
class RBNode<K extends Comparable, V> {
  RBNode<K, V>? _left;
  RBNode<K, V>? _right;
  RBNode<K, V>? _parent;

  /// [left] node.
  RBNode<K, V>? get left => _left;

  /// [right] node.
  RBNode<K, V>? get right => _right;

  /// [parent] node, null for the root node.
  RBNode<K, V>? get parent => _parent;

  K _key;

  /// [key] is a read-only node's field by which the tree is sorted into a binary tree structure.
  K get key => _key;

  /// [value] is a modifiable field associated with every node.
  V value;

  bool _red = true;

  /// [red] flag.
  bool get red => _red;

  RBNode(K key, V value)
      : _key = key,
        value = value;

  /// [isRoot] equals to parent == null.
  bool get isRoot => _parent == null;

  /// [isLeaf] returns true if the node has no children.
  bool get isLeaf => _left == null && _right == null;
}

/// [InsertNodeResult] consists of two fields to describe a result of calling [RBTree.insertNode].
class InsertNodeResult<K extends Comparable, V> {
  RBNode<K, V> _inserted;

  /// [inserted] represents the node that has been added to the tree during [RBTree.insert]. If it replaced another node, the replaced one is returned as [detached].
  RBNode<K, V> get inserted => _inserted;
  RBNode<K, V>? _detached;

  /// [detached] represents the node that has been detached from the tree during [RBTree.insert], if it had the same key as newly added.
  RBNode<K, V>? get detached => _detached;

  InsertNodeResult(this._inserted, this._detached);
}

/// RBTree is a red-black tree binary search tree based on the [K] as key type, which should extend Comparable.
///
/// If the same key appears twice, it overwrites the node and returns a possible disconnected node via [InsertNodeResult] tuple.
///
/// http://www.cs.cornell.edu/courses/cs312/2004fa/lectures/lecture11.htm
///
/// Thank you Franck Bui-Huu for the C implementation.
/// https://github.com/fbuihuu/libtree/blob/master/rb.c
///
class RBTree<K extends Comparable, V> {
  RBNode<K, V>? _root;

  // [root] returns the root node.
  RBNode<K, V>? get root => _root;

  int _length = 0;

  // [length] returns the amount of nodes.
  int get length => _length;

  RBNode<K, V>? _first;

  // [first] returns cached node with a minimum key.
  RBNode<K, V>? get first => _first;

  RBNode<K, V>? _last;

  // [first] returns cached node with a maximum key.
  RBNode<K, V>? get last => _last;

  /// insert returns a [InsertNodeResult], consisting of the just [inserted] node and a possibly [detached] from the tree node
  /// if replacement had to happened.
  InsertNodeResult<K, V> insert(K key, V value) {
    var node = RBNode(key, value);
    return InsertNodeResult(node, insertNode(node));
  }

  /// find for the [key] in the tree. Provide [from] to search inside a subtree.
  RBNode<K, V>? find(K key, {RBNode<K, V>? from}) {
    if (from == null) from = _root;
    RBNode<K, V>? node = from;
    while (node != null && node._key != key) {
      if (key.compareTo(node._key) < 0)
        node = node._left;
      else
        node = node._right;
    }
    return node;
  }

  /// [] is a syntax sugar for [find].
  RBNode<K, V>? operator [](K key) {
    return find(key);
  }

  /// []= adds a node or replaces [RBNode] if an existing [key] is used.
  void operator []=(K key, V value) {
    insert(key, value);
  }

  /// min returns the node with a minimum key starting with the [from] subtree.
  ///
  /// This may result in [from] itself, if it has no left branch node.
  RBNode<K, V> min(RBNode<K, V> from) {
    while (from._left != null) {
      from = from._left!;
    }
    return from;
  }

  /// max returns the node with a maximum key starting with the [from] subtree.
  ///
  /// This may result in [from] itself, if it has no right branch node.
  RBNode<K, V> max(RBNode<K, V> from) {
    while (from._right != null) {
      from = from._right!;
    }
    return from;
  }

  /// predecessor returns the node which key is the biggest of all the smaller than [node.key].
  RBNode<K, V>? predecessor(RBNode<K, V> node) {
    if (node._left != null) return max(node._left!);
    var p = node._parent;
    while (p != null && p._left == node) {
      node = p;
      p = p._parent;
    }
    return p;
  }

  /// successor returns the node which key is the smallest of all the greater than [node.key].
  RBNode<K, V>? successor(RBNode<K, V> node) {
    if (node._right != null) return min(node._right!);
    var p = node._parent;
    while (p != null && p._right == node) {
      node = p;
      p = p._parent;
    }
    return p;
  }

  _leftRotate(RBNode<K, V> node) {
    assert(node._right != null);
    var q = node._right!; // rotate left assumes right side is not null
    var parent = node._parent;
    if (parent == null) {
      _root = q;
    } else {
      if (parent._left == node) {
        parent._left = q;
      } else {
        parent._right = q;
      }
    }
    q._parent = parent;
    node._parent = q;
    node._right = q._left;
    if (node._right != null) node._right!._parent = node;
    q._left = node;
  }

  _rightRotate(RBNode<K, V> node) {
    assert(node._left != null);
    var q = node._left!; // rotate right assumes left side is not null
    var parent = node._parent;
    if (parent == null) {
      _root = q;
    } else {
      if (parent._left == node) {
        parent._left = q;
      } else {
        parent._right = q;
      }
    }
    q._parent = parent;
    node._parent = q;
    node._left = q._right;
    if (node._left != null) node._left!._parent = node;
    q._right = node;
  }

  /// insertNode will move the [node] from any current trees it may be at, thus destroying the other tree's integrity.
  ///
  /// insertNode implants the [node] into the tree at the right location and performs balancing.
  ///
  /// insertNode returns a possibly replaced node, which will be detached from the tree by nullifying its [RBNode.parent], [RBNode.left] and [RBNode.right].
  /// insertMode of a node that is already in the tree is a no-op and returns `null`.
  RBNode<K, V>? insertNode(RBNode<K, V> node) {
    RBNode<K, V>? found = root;
    RBNode<K, V>? parent;
    bool leftSide = false;
    while (found != null) {
      var cmp = node._key.compareTo(found._key);
      if (cmp == 0) {
        break;
      }
      parent = found;
      leftSide = cmp < 0;
      if (leftSide) {
        found = found._left;
      } else {
        found = found._right;
      }
    }
    if (found != null) {
      if (found == node) {
        return null;
      }
      // substituting the node
      var oldParent = found._parent;
      node._parent = oldParent;
      if (oldParent == null) {
        _root = node;
      } else {
        if (oldParent._left == found) {
          oldParent._left = node;
        } else {
          oldParent._right = node;
        }
      }
      if (found._left != null) found._left!._parent = node;
      if (found._right != null) found._right!._parent = node;
      node._red = found._red;
      node._left = found._left;
      node._right = found._right;
      if (_first == found) _first = node;
      if (_last == found) _last = node;
      found._parent = null;
      found._left = null;
      found._right = null;
      return found; // detached old node
    }
    node._left = null;
    node._right = null;
    node._red = true;
    node._parent = parent;
    if (parent == null) {
      _root = node;
      _first = node;
      _last = node;
    } else {
      if (leftSide) {
        if (parent == _first) _first = node;
      } else {
        if (parent == _last) _last = node;
      }
      if (leftSide) {
        parent._left = node;
      } else {
        parent._right = node;
      }
    }
    parent = node._parent;
    while (parent != null && parent._red) {
      var grandpa = parent._parent!;
      if (parent == grandpa._left) {
        var uncle = grandpa._right;
        if (uncle != null && uncle._red) {
          parent._red = false;
          uncle._red = false;
          grandpa._red = true;
          node = grandpa;
        } else {
          if (node == parent._right) {
            _leftRotate(parent);
            node = parent;
            parent = node._parent;
          }
          parent!._red = false;
          grandpa._red = true;
          _rightRotate(grandpa);
        }
      } else {
        var uncle = grandpa._left;
        if (uncle != null && uncle._red) {
          parent._red = false;
          uncle._red = false;
          grandpa._red = true;
          node = grandpa;
        } else {
          if (node == parent._left) {
            _rightRotate(parent);
            node = parent;
            parent = node._parent;
          }
          parent!._red = false;
          grandpa._red = true;
          _leftRotate(grandpa);
        }
      }
      parent = node._parent;
    }
    root!._red = false;
    _length++;
    return null;
  }

  /// removes removes a node with the [key]. It's a no-op if the node doesn't exist.
  remove(K key) {
    var node = find(key);
    if (node == null) return;
    var parent = node._parent;
    var left = node._left;
    var right = node._right;
    RBNode<K, V>? next;
    if (node == _first) _first = successor(node);
    if (node == _last) _last = predecessor(node);
    if (left == null) {
      next = right;
    } else if (right == null) {
      next = left;
    } else {
      next = min(right);
    }
    if (parent != null) {
      if (parent._left == node) {
        parent._left = next;
      } else {
        parent._right = next;
      }
    } else {
      _root = next;
    }
    _length--;
    bool r;
    if (left != null && right != null) {
      r = next!._red;
      next._red = node._red;
      next._left = left;
      left._parent = next;
      if (next != right) {
        parent = next._parent;
        next._parent = node._parent;
        node = next._right;
        parent!._left = node;
        next._right = right;
        right._parent = next;
      } else {
        next._parent = parent;
        parent = next;
        node = next._right;
      }
    } else {
      r = node._red;
      node = next;
    }
    if (node != null) {
      node._parent = parent;
    }
    if (r) {
      return;
    }
    if (node != null && node._red) {
      node._red = false;
      return;
    }
    do {
      if (node == root) break;
      if (node == parent!._left) {
        var sibling = parent._right;
        if (sibling!._red) {
          sibling._red = false;
          parent._red = true;
          _leftRotate(parent);
          sibling = parent._right;
        }
        if (sibling != null &&
            ((sibling._left == null || !sibling._left!._red) &&
                (sibling._right == null || !sibling._right!._red))) {
          sibling._red = true;
          node = parent;
          parent = parent._parent;
          continue;
        }
        if (sibling!._right == null || !sibling._right!._red) {
          sibling._left!._red = false;
          sibling._red = true;
          _rightRotate(sibling);
          sibling = parent._right;
        }
        sibling!._red = parent._red;
        parent._red = false;
        sibling._right!._red = false;
        _leftRotate(parent);
        node = root;
        break;
      } else {
        var sibling = parent._left;
        if (sibling!._red) {
          sibling._red = false;
          parent._red = true;
          _rightRotate(parent);
          sibling = parent._left;
        }
        if (sibling != null &&
            ((sibling._left == null || !sibling._left!._red) &&
                (sibling._right == null || !sibling._right!._red))) {
          sibling._red = true;
          node = parent;
          parent = parent._parent;
          continue;
        }
        if (sibling!._left == null || !sibling._left!._red) {
          sibling._right!._red = false;
          sibling._red = true;
          _leftRotate(sibling);
          sibling = parent._left;
        }
        sibling!._red = parent._red;
        parent._red = false;
        sibling._left!._red = false;
        _rightRotate(parent);
        node = root;
        break;
      }
    } while (!node._red);
    node?._red = false;
  }

  /// toString will act as `output(OutputStyle.KeyDesc).toString()`.
  @override
  String toString() {
    return output(OutputStyle.KeyDesc).toString();
  }

  /// output returns a [StringBuilder] and allows to specify how to output the tree.
  StringBuffer output(OutputStyle style) {
    var b = StringBuffer();
    if (style == OutputStyle.Graphviz) {
      b.write("digraph {");
    }
    _convert(_root, b, style, _PrimitiveWrapper<int>(0), 0);
    if (style == OutputStyle.Graphviz) {
      b.write("\n}");
    }
    return b;
  }

  _convert(RBNode<K, V>? n, StringBuffer b, OutputStyle style,
      _PrimitiveWrapper<int> idx, int side) {
    if (n == null) {
      return;
    }
    switch (style) {
      case OutputStyle.KeyDesc:
        b.write("(${n._key}");
        if (side != 0) {
          b.write(side < 0 ? "<" : ">");
          b.write(n._red ? "r" : "b");
        } // avoid any output for root
        if (n._left != null) _convert(n._left, b, style, idx, -1);
        if (n._right != null) _convert(n._right, b, style, idx, 1);
        b.write(")");
        break;
      case OutputStyle.KeyList:
        if (b.length > 0) b.write(",");
        b.write("${n._key}");
        if (n._left != null) _convert(n._left, b, style, idx, -1);
        if (n._right != null) _convert(n._right, b, style, idx, 1);
        break;
      case OutputStyle.Graphviz:
        if (n._parent != null) {
          if (b.length > 0) {
            b.write("\n");
          }
          b.write("  ${n._parent!._key.toString()} -> ${n._key.toString()};");
        }
        // style is needed for the root node too
        _writeStyle(b, n._key.toString(), n._red ? "tomato" : "black");
        if (n._left == null) {
          _writeNull(b, idx, n);
        } else {
          _convert(n._left, b, style, idx, 1);
        }
        if (n._right == null) {
          _writeNull(b, idx, n);
        } else {
          _convert(n._right, b, style, idx, -1);
        }
    }
  }

  void _writeStyle(StringBuffer b, String name, String color) {
    if (b.length > 0) {
      b.write("\n");
    }
    b.write("  $name[style=filled,color=$color,fontcolor=white];");
  }

  void _writeNull(StringBuffer b, _PrimitiveWrapper<int> idx, RBNode<K, V> n) {
    if (b.length > 0) {
      b.write("\n");
    }
    b.write("  null$idx [shape=point];\n");
    b.write("  ${n._key.toString()} -> null$idx;");
    idx._value++;
  }

  /// forEach visits every node starting with the [min] and moves to [successor] and
  /// calls a callback for each one of the node with an option to interrupt the
  /// flow by returning [false] out of it.
  forEach(bool Function(RBNode<K, V> node) callback) {
    if (root == null) return;
    RBNode<K, V>? node = min(root!);
    while (node != null) {
      if (callback(node)) {
        node = successor(node);
      } else {
        break;
      }
    }
  }

  /// [assertConformity] can be used in tests to check whether the tree conforms to all the red-black tree rules:
  ///
  /// 1. Every node is either red or black (provided by [red] boolean)
  /// 2. The root is black.
  /// 3. Every null-leaf is black (provided by design).
  /// 4. If a node is red, then both its children are black (in other words, a red child can't have a red parent).
  /// 5. For each node, all simple paths from the node to descendant leaves contain the
  /// same number of black nodes.
  assertConformity() {
    if (root != null) assert(root!._red == false); // #2
    var cnt = 0;
    forEach((v) {
      if (v.isLeaf) {
        var count = 0;
        RBNode<K, V>? s = v;
        while (s != null) {
          if (s.red) {
            if (s._parent != null) assert(!s._parent!._red); // #4
          } else {
            count++; // #5 amount of blacks in a leaf
          }
          s = s.parent;
        }
        if (cnt == 0) cnt = count;
        assert(count == cnt);
      }
      return true;
    });
  }

  /// randomNode returns a random node based on the provided seed.
  RBNode<K, V>? randomNode(Random rnd) {
    if (_root == null) return null;
    RBNode<K, V>? toDelete;
    var idx = rnd.nextInt(_length);
    var i = 0;
    forEach((node) {
      if (idx == i) {
        toDelete = node;
        return false;
      }
      i++;
      return true;
    });
    return toDelete;
  }
}

/// OutputStyle is a enum for possible output styles.
enum OutputStyle {
  /// Outputs as something like `(50.b(45<r(25<b)(48>b(46<r)))(150>r(60<b)(180>b(190>r))))`.
  KeyDesc,

  /// Outputs as comma-separated keys.
  KeyList,

  /// Outputs as `digraph { ... }`.
  Graphviz,
}

class _PrimitiveWrapper<T> {
  T _value;
  _PrimitiveWrapper(this._value);
  @override
  String toString() {
    return _value.toString();
  }
}


example/tree_structures_example.dart
import 'package:tree_structures/tree_structures.dart';

main() {
  var tree = RBTree<int, dynamic>();
  tree.insert(100, "data associated with 100");
  tree.insert(150, "data associated with 150");
  tree.remove(100);
  tree.insert(200, "data associated with 200");
  tree.insert(20, "data associated with 20");
  tree.insert(25, "data associated with 25");
  tree.insert(18, "data associated with 18");
  tree.insert(120, "data associated with 120");
  tree.insert(180, "data associated with 180");
  print(tree);
  tree.forEach((node) {
    print("${node.key}:${node.value}");
    return true; // continue
  });
  print(tree.output(OutputStyle.Graphviz));
  var sb = StringBuffer();
  tree.forEach((node) {
    if (sb.length > 0) {
      sb.write(",");
    }
    sb.write("${node.key}");
    return true;
  });
  print(sb.toString());
}
